server/region_handler.go

```diff
@@ -0,0 +1,493 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/domain"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/codec"
+	goctx "golang.org/x/net/context"
+)
+
+// RegionsHTTPRequest is for regions http request.
+type RegionsHTTPRequest struct {
+	resp        http.ResponseWriter
+	req         *http.Request
+	server      *Server
+	params      map[string]string
+	reqID       string
+	regionCache *tikv.RegionCache
+}
+
+// NewRegionsHTTPRequest create a RegionsHTTPRequest
+func NewRegionsHTTPRequest(s *Server, w http.ResponseWriter, req *http.Request, regionCache *tikv.RegionCache) *RegionsHTTPRequest {
+	regionCache.EmptyRegions()
+	return &RegionsHTTPRequest{
+		server:      s,
+		req:         req,
+		resp:        w,
+		params:      mux.Vars(req),
+		regionCache: regionCache,
+		//TODO genuuid
+		reqID: fmt.Sprintf("http_%d", time.Now().UnixNano()),
+	}
+}
+
+func (req *RegionsHTTPRequest) writeResponse(data interface{}, err error) {
+	var retItem *RegionsResponse
+	if err != nil {
+		retItem = NewRegionsResponseWithError(req.reqID, err)
+	} else {
+		retItem = NewRegionsResponseWithData(req.reqID, data)
+	}
+	js, err := json.Marshal(retItem)
+	if err != nil {
+		req.resp.WriteHeader(http.StatusInternalServerError)
+		req.resp.Write([]byte(err.Error()))
+		return
+	}
+
+	req.resp.Header().Set("Content-Type", "application/json")
+	if !retItem.Success {
+		req.resp.WriteHeader(http.StatusBadRequest)
+	} else {
+		req.resp.WriteHeader(http.StatusOK)
+	}
+	req.resp.Write(js)
+}
+
+// HandleListRegions handles http request for list regions.
+func (req *RegionsHTTPRequest) HandleListRegions() {
+	dbName, _ := req.params["db"]
+	tableName, _ := req.params["table"]
+	data, err := req.listTableRegions(dbName, tableName)
+	req.writeResponse(data, err)
+}
+
+// HandleGetRegionByID handles request for get region by ID
+func (req *RegionsHTTPRequest) HandleGetRegionByID() {
+	regionIDStr, _ := req.params["regionID"]
+	regionID, err := strconv.ParseInt(regionIDStr, 0, 64)
+	if err != nil {
+		req.writeResponse(nil, err)
+		return
+	}
+	data, err := req.getRegionWithRegionID(uint64(regionID))
+	req.writeResponse(data, err)
+}
+
+// prepare creates a Backoffer,Domain when store is tikv.
+func (req *RegionsHTTPRequest) prepare() (bo *tikv.Backoffer, domain *domain.Domain, err error) {
```

Comment:
If we always have to call `prepare()` before using `RegionHTTPRequest`, it should be better to prepare in the new function.